File: algs4_java/algs4-master-RQTN/algs4-master/src/stacks_and_queues/zhr_code/LinkedStack.java
import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class LinkedStack<Item> implements Stack<Item>{
    // list node
    private class Node{
        Item Item;
        Node next;
    }
    private Node first; // 栈顶元素
    private int N; // 元素数量

    // public Stack<Item>(){ // 构造函数
    //     first = new Node();
    //     first.next = null;
    //     N = 0;
    // }

    public void push(Item Item){// 添加元素
        Node node = new Node();
        node.Item = Item;
        node.next = first;
        first = node;
        N++;
    }
    public Item pop(){
        Item Item = first.Item;
        Node tmp = first;
        first = first.next;
        tmp.next = null; // help GC
        N--;
        return Item;
    } // 删除元素
    public boolean isEmpty(){
        return N == 0;   // or return first == null;
    } // 是否为空
    public int size(){
        return N;
    } // 元素数量
    // <Item> stack = new LinkedStack<>(); // 创建栈对象

    public static void main(String[] args) {
        LinkedStack<String> stack = new LinkedStack<String>();
        while (!StdIn.isEmpty()) {
            
            String s = StdIn.readString();
            if (s.equals("-")) {
                stack.pop();
            }
            else {
                stack.push(s);
            }
            // 手动遍历
            StdOut.printf("%4s | Stack Display: ", s);
            LinkedStack<String>.Node cur = stack.first;
            while (cur != null) {
                StdOut.print(cur.Item + " ");
                cur = cur.next; 
            }
            StdOut.println(); // 换行
        }
        StdOut.println("(" + stack.size() + " left on stack)");
        // 遍历打印栈里的内容
        // for (String s : stack) {
        //     StdOut.println(s);
        // }  // 需要实现Iterable接口



    }
}

File: algs4_cpluspuls/stacks_and_queues/src/LinkedStack.cpp
#include "Stack.h"
#include <iostream>
#include <string>

template <typename T>
class LinkedStack: public Stack<T>
{
public:
    struct Node
    {
        T item;
        Node* next;
        Node(): item(T()), next(nullptr) {} // 默认构造函数
    };
public:
    Node* first; // top of stack
    int n;       // size of the stack

public:
    LinkedStack():n(0),first(nullptr) {}         // create an empty stack
    ~LinkedStack() override{
        while (!isEmpty()) {
            pop();
        }
    }         // destroy the stack


    void push(T item) override{
        Node* oldFirst = first;
        first = new Node;
        first->item = item;
        first->next = oldFirst;
        n++;
    }     // add an item to the top of the stack
    T pop() override{
        if (isEmpty()) {
            throw std::runtime_error("Stack underflow");
        }
        T item = first->item;
        Node* oldFirst = first;
        first = first->next;
        delete oldFirst;
        n--;
        return item;
    }               // remove and return the item from the top of the stack
    bool isEmpty() override{
        return first == nullptr;
    }        // is the stack empty?
    int size(){
        return n;
    }             // return the number of items on the stack
};


int main(int argc, char* argv[]) {
    LinkedStack<std::string> stack;
    std::string s;
    while (std::cin >> s) {
        if (s == "-") {
            if (!stack.isEmpty()) {
                stack.pop();
            }
        } else {
            stack.push(s);
        }
        // 手动遍历
        std::cout  << s << " | Stack Display: ";
        LinkedStack<std::string>::Node* cur = stack.first;
        while (cur != nullptr) {
            std::cout << cur->item << " ";
            cur = cur->next;
        }
        std::cout << std::endl; // 换行
    }
    std::cout << "(" << stack.size() << " left on stack)" << std::endl;
    return 0;
}